### PR TITLE
Refactoring rendering system

### PR DESCRIPTION
**Pull request by nashmit**
The rendering system was refactored in order to accept sub-meshes as input from CGAL meshes.

### CODE REVIEW

**Review by lkeegan — COMMENTED** [review]
Thanks - looks good!

I struggle a bit with the user interface, e.g. I can't look at the other side of the mesh from the initial viewpoint. A suggestion for the mouse interaction would be to offer the same as vtk does in the voxel renderer, where the camera doesn't move, click+drag rotates the entire mesh about the centre, and the mouse wheel zooms in/out. I think you already implemented (but don't yet use) a rotation?

File: gui/widgets/qopenglmousetracker.hpp [hunk 1]
```diff
@@ -28,7 +28,16 @@ class QOpenGLMouseTracker : public QOpenGLWidget {
...
+  /**
+   * @brief Number of visible submeshes is influenced by the number of colors.
+   *
+   * The colors are distributed to submeshes in the same order they are
+   * provided. In case of a smaller number of colors, submeshes with higher
+   * index will not be visible.
+   */
+
+  void SetSubMeshes(const sme::mesh::Mesh3d &mesh,
+                    const std::vector<QColor> &colors = std::vector<QColor>(0));
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.hpp:40]: Here it would be nice to by default use the colours from the mesh unless otherwise specified

**Review by lkeegan — COMMENTED** [review]
Also a minor thing: it would be nice to use the same background colour as the other widgets, here is how to get that colour from qt: https://github.com/spatial-model-editor/spatial-model-editor/blob/368da9e6cb513310f47e976ab6608e0fe8feebda/gui/widgets/qvoxelrenderer.cpp#L31-L34

**Comment by nashmit** [comment]
> Also a minor thing: it would be nice to use the same background colour as the other widgets, here is how to get that colour from qt:
> 
> https://github.com/spatial-model-editor/spatial-model-editor/blob/368da9e6cb513310f47e976ab6608e0fe8feebda/gui/widgets/qvoxelrenderer.cpp#L31-L34

done.

File: core/mesh/src/mesh3d.cpp [hunk 2]
```diff
@@ -381,4 +386,14 @@ QString Mesh3d::getGMSH() const {
   return msh;
 }
 
+std::vector<QColor> Mesh3d::getColorTable() const { return colorTableVec; }
+
+QVector3D Mesh3d::getOffset() const {
+
+  return QVector3D(
+      -static_cast<float>(image3_->image()->vx * image3_->xdim() / 2.0f),
```
Comment (lkeegan) [inline core/mesh/src/mesh3d.cpp:394]: don't you also need to subtract the overall offset here (which I guess is 0 in your tests)?
i.e. `-image3_->image()->tx` etc

File: gui/widgets/qopenglmousetracker.cpp [hunk 3]
```diff
@@ -12,7 +12,18 @@ QOpenGLMouseTracker::QOpenGLMouseTracker(float lineWidth,
     : m_camera(cameraFOV, static_cast<float>(size().width()),
                static_cast<float>(size().height()), cameraNearZ, cameraFarZ),
       m_lineWidth(lineWidth), m_lineSelectPrecision(lineSelectPrecision),
-      m_selectedObjectColor(selectedObjectColor), m_frameRate(frameRate) {}
+      m_selectedObjectColor(selectedObjectColor), m_frameRate(frameRate) {
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:15]: I think you should be able to just add this to the initialiser list:
`backgroundColor(palette().color(QWidget::backgroundRole()))`
Also shouldn't it be `m_backgroundColor` to match your naming convention here?

File: gui/widgets/qopenglmousetracker.cpp [hunk 4]
```diff
@@ -217,11 +237,33 @@ QVector3D QOpenGLMouseTracker::GetCameraOrientation() const {
...
+
 void QOpenGLMouseTracker::SetSubMeshes(const sme::mesh::Mesh3d &mesh,
                                        const std::vector<QColor> &colors) {
 
+  if (colors.size() == 0) {
+    m_SubMeshes = std::unique_ptr<rendering::WireframeObjects>(
+        new rendering::WireframeObjects(mesh, this, mesh.getColorTable(),
+                                        mesh.getOffset()));
+  }
+
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker.cpp:264]: should there be an `else` here? (or equivalently a `return` above)

File: gui/widgets/qopenglmousetracker_t.cpp [hunk 5]
```diff
@@ -31,11 +31,11 @@ TEST_CASE("QOpenGLMouseTracker: OpenGL", tags) {
   QColor greenColor = QColor(0, 255, 0);
   QColor yelloColor = QColor(255, 255, 0);
 
+  test.setBackground(blackColor);
   test.show();
-  //  wait(100);
 
   // camera position
```
Comment (lkeegan) [inline gui/widgets/qopenglmousetracker_t.cpp:37]: it would be nice to have a test for the rotation / offset / default colours

**Review by lkeegan — APPROVED** [review]
Looks good - I'll merge this then I can open a PR with it integrated in the GUI & remaining issues can be a separate PR

**Pull request merged**